### ISSUE TITLE
(Question) Bearer authentication with Angular SSR not working

### ISSUE THREAD

**Issue by FreedCapybara** [aspnet/JavaScriptServices#1308]
I've implemented the basic ideas from [this issue](https://github.com/angular/universal-starter/issues/373), but somehow I always get `Microsoft.AspNetCore.NodeServices:Error: Unhandled Promise rejection: Converting circular structure to JSON` when trying to hit any endpoint with an `[Authorize]` attribute during SSR.

My app uses JWT authentication with the token stored in a cookie so I can access it on the server. I know the authorization header is getting set correctly on the server during SSR because I can `console.log` the token and see it in the server output. I tried enabling CORS in case the Node stuff is running on a separate port, but that hasn't made a difference. The exact same request code works perfectly when run on the client.

What am I missing? 

Thanks in advance!

**Comment by SteveSandersonMS** [comment]
I don't know why you're getting that error, but it fairly strongly implies that some of the data you're trying to serialize somewhere contains a cyclic reference.

My best guess is that your backend isn't in fact returning the data you think it is returning, but instead is returning something like a redirection to a login page, which itself is not serializable. You might want to check in more detail that you are really setting the auth context correctly and the backend is really returning the data you think.

I'll close this as it sounds like an application-specific issue rather than an issue in any code in this repo. Hope that's OK!

**Comment by FreedCapybara** [comment]
Thanks Steve. I'm guessing it's because an exception is getting thrown and it's trying to serialize the exception. Unfortunately nothing shows in Visual Studio and the zonejs error output isn't helpful, so I'm not sure how to debug the issue.

Do you know of any documentation or working examples of SSR with authentication that I could look at? There are a few StackOverflow questions and several issues in various repos, but it's hard to know what final solution ended up working.

Thanks!
________________________________

**Comment by FreedCapybara** [comment]
If it helps anything, here's my implementation so far:

**Views/Home/Index.cshtml** passes cookies to Node
```html
<app asp-prerender-module="ClientApp/dist/main-server"
     asp-prerender-data="new { Cookies = ViewContext.HttpContext.Request.Cookies }">
    Loading...
</app>
```

**boot-server.ts:** polyfills cookies for use with [ngx-cookie](https://github.com/salemdar/ngx-cookie)
```js
export default createServerRenderer(params => {
  var cookies = ''; // cookies come in as [{ key, value}] -- turn them into a normal cookie string
  for (let cookie of params.data.cookies) {
    cookies += `${cookie.key}=${cookie.value}; `;
  }
  document.cookie = cookies; // <-- polyfilled to work with ngx-cookie

  // ...
}
```

**HttpService.ts** is a wrapper service for Angular's Http to control adding headers, etc.
```js
// GET requests are made like this
var headers = new Headers();
var token = this.cookieService.get('token');
if (token) {
  headers.append('Authorization', `Bearer ${token}`);
}
this.http.get(this.baseUrl + url, { headers: headers }) // (this.http is Angular's standard Http)
  .subscribe( ... );
```

Using the Node debugger [as outlined here](https://github.com/aspnet/JavaScriptServices/tree/dev/src/Microsoft.AspNetCore.SpaServices#debugging-your-javascripttypescript-code-when-it-runs-on-the-server), it's clear that the header is being set correctly. Setting a breakpoint and inspecting the `errorInfo` callback argument in boot-server.ts shows this.

**Startup.cs** contains the bearer authentication setup
```c#
        public IServiceProvider ConfigureServices(IServiceCollection services)
        {
            services.AddIdentity<ApplicationUser, IdentityRole>(options =>
            {
                options.Password.RequireDigit = false;
                options.Password.RequireLowercase = false;
                options.Password.RequireNonAlphanumeric = false;
                options.Password.RequireUppercase = false;
                options.Password.RequiredLength = 6;

                options.Lockout.DefaultLockoutTimeSpan = TimeSpan.FromMinutes(30);
                options.Lockout.MaxFailedAccessAttempts = 100;

                options.User.RequireUniqueEmail = true;
            })
            .AddEntityFrameworkStores<ApplicationDbContext>()
            .AddDefaultTokenProviders();

            services.AddAuthentication(options =>
            {
                options.DefaultAuthenticateScheme = JwtBearerDefaults.AuthenticationScheme;
                options.DefaultChallengeScheme = JwtBearerDefaults.AuthenticationScheme;
            })
            .AddJwtBearer(options =>
            {
                options.TokenValidationParameters = new TokenValidationParameters
                {
                    IssuerSigningKey = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(Configuration.GetSection("Environment:TokenSigningKey").Value)),
                    ValidateIssuer = true,
                    ValidIssuer = Configuration.GetSection("Environment:BaseUrl").Value,
                    ValidateAudience = true,
                    ValidAudience = Configuration.GetSection("Environment:BaseUrl").Value,
                    ValidateLifetime = true,
                    ClockSkew = TimeSpan.Zero
                };
                options.RequireHttpsMetadata = false;
            });

            // ...
        }

        public void Configure(IApplicationBuilder app, IHostingEnvironment env, IServiceProvider serviceProvider)
        {
            // ...
            app.UseAuthentication();
            // ...
        }
```

Everything works like a dream until trying to load stuff with SSR.

**Comment by FreedCapybara** [comment]
After some fiddling, it looks like all HTTP requests fail if performed during SSR, even using the base Angular Http with no authorization header to a non-guarded endpoint.

**Comment by FreedCapybara** [comment]
To keep documenting the process, it looks like it might be related to SSL. Trying to find where the http request fails in the stack trace (to find out what is causing the circular json error) I see `Error: unable to verify the first certificate`.

Haven't been able to fix it yet 😕

**Comment by FreedCapybara** [comment]
@SteveSandersonMS could you tell me what is the Angular equivalent of your fix in #1089? I think it might finally solve my issue 🎉

I tried passing `options.agent` to the http request like this, but haven't gotten it to work:
```js
    if (isPlatformServer(this.platformId)) {
      let Agent = (require('https') as any).Agent;
      options.agent = new Agent({ rejectUnauthorized: options.useHttpsValidation }) as RequestInit;
    }

    // ...

    this.http.get(this.baseUrl + url, options); // this.http is just the basic Http service
```

Thanks in advance!

**Comment by SteveSandersonMS** [comment]
@FreedCapybara I don't know whether Angular's `http` API exposes a method for disabling cert verification - please see Angular docs for details about that.
